### PR TITLE
[Enhancement](6/n) Support modify column type without data rewriting in share data mode

### PR DESCRIPTION
**Pull request by sduzh**
This PR is part of the PR stack to support fast schema evolution in share data mode.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/rowset/column_iterator.cpp [hunk 1]
```diff
@@ -64,4 +64,28 @@ Status ColumnIterator::fetch_dict_codes_by_rowid(const Column& rowids, Column* v
     return fetch_dict_codes_by_rowid(p, rowids.size(), values);
 }
 
+Status ColumnIterator::next_batch(const SparseRange<>& range, Column* dst) {
```
Comment (sduzh) [inline be/src/storage/rowset/column_iterator.cpp:67]: Just added a default implementation of `next_batch`, no logical modification.

File: be/src/storage/rowset/column_iterator.cpp [hunk 2]
```diff
@@ -64,4 +64,28 @@ Status ColumnIterator::fetch_dict_codes_by_rowid(const Column& rowids, Column* v
...
+        auto r = Range<>{iter.next(to_read)};
+        auto n = size_t{r.span_size()};
+        RETURN_IF_ERROR(next_batch(&n, dst));
+        CHECK_EQ(r.span_size(), n);
+        to_read -= n;
+    }
+    return Status::OK();
+}
+
+Status ColumnIterator::fetch_values_by_rowid(const rowid_t* rowids, size_t size, Column* values) {
```
Comment (sduzh) [inline be/src/storage/rowset/column_iterator.cpp:81]: Added a default implementation of `fetch_values_by_rowid`, the original implementation will return `NotSupported`

File: be/src/storage/rowset/column_iterator.h [hunk 3]
```diff
@@ -155,13 +154,14 @@ class ColumnIterator {
...
 
-    /// for vectorized engine
-    [[nodiscard]] virtual Status get_row_ranges_by_zone_map(const std::vector<const ColumnPredicate*>& predicates,
-                                                            const ColumnPredicate* del_predicate,
-                                                            SparseRange<>* row_ranges) = 0;
+    virtual Status get_row_ranges_by_zone_map(const std::vector<const ColumnPredicate*>& predicates,
+                                              const ColumnPredicate* del_predicate, SparseRange<>* row_ranges) {
+        row_ranges->add({0, static_cast<rowid_t>(num_rows())});
+        return Status::OK();
+    }
```
Comment (sduzh) [inline be/src/storage/rowset/column_iterator.h:161]: Added a default implementation to reduce code duplication

File: be/src/storage/rowset/column_iterator.h [hunk 4]
```diff
@@ -189,17 +189,15 @@ class ColumnIterator {
     // batch of dictionary codes for dictionary encoded values.
     // this method can be invoked only if `all_page_dict_encoded` returns true.
     // type of |dst| must be `FixedLengthColumn<int32_t>` or `NullableColumn(FixedLengthColumn<int32_t>)`.
-    [[nodiscard]] virtual Status next_dict_codes(size_t* n, Column* dst) { return Status::NotSupported(""); }
```
Comment (sduzh) [inline be/src/storage/rowset/column_iterator.h:192]: `[[nodiscard]]` declaration for virtual function is useless so I removed it.

File: be/src/storage/rowset/column_reader.cpp [hunk 5]
```diff
@@ -518,12 +518,20 @@ Status ColumnReader::_zone_map_filter(const std::vector<const ColumnPredicate*>&
...
                                       std::unordered_set<uint32_t>* del_partial_filtered_pages,
                                       std::vector<uint32_t>* pages) {
+    LogicalType lt;
+    if (!predicates.empty()) {
+        lt = predicates[0]->type_info()->type();
+    } else if (del_predicate) {
+        lt = del_predicate->type_info()->type();
+    } else {
+        return Status::OK();
+    }
```
Comment (sduzh) [inline be/src/storage/rowset/column_reader.cpp:528]: The type of the predicate may be different from the data type in the segment file, so here it is changed to use the type of the predicate to parse the zone map string

File: be/src/storage/rowset/dictcode_column_iterator.h [hunk 6]
```diff
@@ -25,134 +25,86 @@
...
-
-    [[nodiscard]] Status decode_dict_codes(const int32_t* codes, size_t size, Column* words) override {
-        return _col_iter->decode_dict_codes(codes, size, words);
-    }
-
-    [[nodiscard]] Status get_row_ranges_by_zone_map(const std::vector<const ColumnPredicate*>& predicates,
-                                                    const ColumnPredicate* del_predicate,
-                                                    SparseRange<>* row_ranges) override {
-        return _col_iter->get_row_ranges_by_zone_map(predicates, del_predicate, row_ranges);
-    }
```
Comment (sduzh) [inline be/src/storage/rowset/dictcode_column_iterator.h:78]: Replaced the implementations here with the default implementations in ColumnIteratorWrapper.

File: fe/fe-core/src/main/java/com/starrocks/catalog/SchemaInfo.java [hunk 7]
```diff
@@ -265,3 +265,4 @@ public SchemaInfo build() {
         }
     }
 }
+
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/catalog/SchemaInfo.java:268]: unnecessary change.

File: fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java [hunk 8]
```diff
@@ -2458,13 +2463,8 @@ public void modifyTableAddOrDropColumns(Database db, OlapTable olapTable,
...
 
-                if (hasMv && indexSchema.size() < originSchema.size()) {
-                    // drop column
-                    List<Column> differences = originSchema.stream().filter(element ->
-                            !indexSchema.contains(element)).collect(Collectors.toList());
-                    // can just drop one column one time, so just one element in differences
-                    int dropIdx = originSchema.indexOf(differences.get(0));
-                    modifiedColumns.add(originSchema.get(dropIdx).getName());
+                if (hasMv) {
+                    modifiedColumns.addAll(AlterHelper.collectDroppedOrModifiedColumns(originSchema, indexSchema));
```
[the remark this answers is not in the document]
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java:2464]: The original one is actually wrong, because it is allowed to add and drop columns at the same time, in which case the number of columns is unchanged.

File: fe/fe-core/src/main/java/com/starrocks/catalog/SchemaInfo.java (hunk 7 above)
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/catalog/SchemaInfo.java]: Removed

File: fe/fe-core/src/main/java/com/starrocks/alter/LakeTableSchemaChangeJob.java [hunk 9]
```diff
@@ -651,30 +651,12 @@ private Set<String> collectModifiedColumnsForRelatedMVs(@NotNull LakeTable tbl)
...
-                // drop column
-                List<Column> differences = originSchema.stream().filter(element ->
-                        !shadowSchema.contains(element)).collect(Collectors.toList());
-                // can just drop one column one time, so just one element in differences
-                int dropIdx = originSchema.indexOf(differences.get(0));
-                modifiedColumns.add(originSchema.get(dropIdx).getName());
-            } else {
-                // add column should not affect old mv, just ignore.
-            }
+            modifiedColumns.addAll(AlterHelper.collectDroppedOrModifiedColumns(originSchema, shadowSchema));
```
[the remark this answers is not in the document]
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/alter/LakeTableSchemaChangeJob.java:659]: FE UT has been added to cover this case.

File: be/src/storage/rowset/cast_column_iterator.h [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+namespace starrocks {
+
+class Column;
+class ColumnRef;
+class Expr;
+class ObjectPool;
+
+class CastColumnIterator : public ColumnIteratorDecorator {
+public:
+    // Will take the ownership of |source_iter|
```
Comment (kevincai) [inline be/src/storage/rowset/cast_column_iterator.h:31]: is it possible to require the `source_iter` to be a unique_ptr, explicitly expressing the acquire of the ownership.

**Review by kevincai — DISMISSED** [review]

**Review by kevincai — DISMISSED** [review]

File: be/src/storage/rowset/cast_column_iterator.h (hunk 10 above)
Comment (sduzh) [reply be/src/storage/rowset/cast_column_iterator.h]: updated

**Review by kevincai — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java [hunk 11]
```diff
@@ -642,7 +644,10 @@ private boolean processModifyColumn(ModifyColumnClause alterClause, OlapTable ol
             }
         } // end for handling other indices
 
-        return false;
+        if (modColumn.isKey() || !modColumn.getType().isScalarType()) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java:647]: why modify column does not support fast schema evolution?

**Review by wyb — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java (hunk 11 above)
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java:647]: Modifying a key column involves the encoding of a short key, which is relatively complex and is not supported for the time being.

**Review by HangyuanLiu — APPROVED** [review]
